### PR TITLE
Add support for Feature Flags.

### LINKED ISSUE #1943: CiviForm lacks a feature flag or experimentation system
When new code is introduced to CiviForm, there is no way to enable or disable it, only deploy or rollback.

We should implement release decoupling for CiviForm using either an open source or custom-made feature flag or experiments system.

- [x] Feature Flag
- [ ] Set feature flags via url (ease of testing in staging etc.)
- [ ] Experiment System

### PR DESCRIPTION
**Pull request by shanemc-goog**
### Description
Adds Injectable annotations and a Module provider that surfaces the system application configuration for features that should be on or off across the entire application.

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #1943

### CODE REVIEW

File: server/conf/application.conf [hunk 1]
```diff
@@ -526,3 +527,7 @@ api_keys_ban_global_subnet = ${?CIVIFORM_API_KEYS_BAN_GLOBAL_SUBNET}
 # load balancer timeout.
 api_applications_list_max_page_size = 1000
 api_applications_list_max_page_size = ${?CIVIFORM_API_APPLICATIONS_LIST_MAX_PAGE_SIZE}
+
+# Feature flag toggles.
+status_tracking_enabled = true
```
Comment (goldblatt) [inline server/conf/application.conf:532]: maybe would default to false here

Comment (shanemc-goog) [reply server/conf/application.conf:532]: totally.  Just playing around atm.  Will fix up and send out formally.

File: server/app/modules/FeatureFlagsModule.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+public class FeatureFlagsModule extends AbstractModule {
+  @Override
+  public void configure() {
+  }
+
+  @Provides
+  @StatusTrackingEnabled
+  public boolean provideStatusTrackingEnabled(Config config) {
+    return checkNotNull(config).getBoolean("status_tracking_enabled");
```
Comment (clouser2) [inline server/app/modules/FeatureFlagsModule.java:18]: This works, but is there a need to add Guice annotations when the config is already injected at the callsite? For example, the flag could be a static String and the callsites could evaluate against the config.

One argument in favor of Guice annotations would be it'd be easier to mock out data in unit tests.

File: server/conf/application.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (shanemc-goog) [reply server/conf/application.conf:532]: yeah, this is just a draft for the general idea.

File: server/app/modules/FeatureFlagsModule.java (hunk 2 above)
Comment (shanemc-goog) [reply server/app/modules/FeatureFlagsModule.java:18]: 1. I think it's good to abstract these mechanics out, especially as we add more Feature Flags and ones that will be long lived.  The values will also be treated more "readably" through the code base as class ctors will generally have to accept the injected value and save it as a class field.
2. I hope we can support url overrides in dev/staging and having that handled in one place is good.
3. Not all involved classes will have the config injected also I think.
4. Testing does seem easier and the "concern" is treated as a first party thing as you mention

Comment (clouser2) [reply server/app/modules/FeatureFlagsModule.java:18]: Definitely seems like reasonable rationale. Another nice thing that would fall out of this is if we wanted to eventually allow admins / specific user types to "force" experiments. Having a single Guice-injection site that can wrap that decision up seems preferable to having each consumer of the feature flag have to apply that logic.

Comment (shanemc-goog) [reply server/app/modules/FeatureFlagsModule.java:19]: yeah, def having wide eyed views into that future :D

**Review by clouser2 — APPROVED** [review]

**Review by bion — APPROVED** [review]

**Pull request merged**